### PR TITLE
Updated README as per latest version

### LINKED ISSUE #15: Detailed documentation in README
- Update the README with all detailed steps to use code runner.

### PR DESCRIPTION
**Pull request by balandi1**
- Added usage of code runner as a base docker image.
- Added instructions to run the language specific docker image.
- Added information about supported languages.
- Updated the web page image.
- Merging of this pull request closes Issue 15.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,21 +3,19 @@ Code Runner: A tool to help run your assignments
 Code Runner is a web application to run code pertaining to your assignment using user provided configurations.
 
 ## Features
-There are two roles in the application - Professor and Student.
-#### Professor
-- Professor provides course specific configurations for creating suitable docker environment.
-- Course specific configurations include specifying operating system and its version, programming language and compiler.
+There is one role in the application - Student.
```
Comment (pradykaushik) [inline README.md]: We can get rid of this now as both the teaching staff and the user have access to the same features.

File: README.md [hunk 2]
```diff
@@ -3,21 +3,19 @@ Code Runner: A tool to help run your assignments
...
 #### Student
```
Comment (pradykaushik) [inline README.md]: we can get rid of this sub-heading for the same reason mentioned above.

File: README.md [hunk 3]
```diff
@@ -3,21 +3,19 @@ Code Runner: A tool to help run your assignments
...
 - Students can submit their assignments and view the output through a web page.
-- They can also enter the optional command line arguments that are required to run their assignment.
+- They can enter the assignment working directory, commands to compile and run the assignment.
```
Comment (pradykaushik) [inline README.md]: as this is a separate point, you shouldn't start the sentence with "They". If you want to do that then it should be a subpoint.

File: README.md [hunk 4]
```diff
@@ -3,21 +3,19 @@ Code Runner: A tool to help run your assignments
...
+- They can also enter the optional command line arguments required to run their assignment.
```
Comment (pradykaushik) [inline README.md]: instead list all the things that can be provided by the user and explain each of them alongside their listing.

File: README.md [hunk 5]
```diff
@@ -3,21 +3,19 @@ Code Runner: A tool to help run your assignments
...
 - Web page accepts assignment files in compressed format. Allowed formats are .zip, .tar and .tar.gz.
       
-Currently the application supports only server features:  
-- Run the server
-- Accept assignment files and command line argument
-- Extract the files send status response to web page.
+Currently the application supports following server features:  
+- Run the server.
```
Comment (pradykaushik) [inline README.md]: how is this a server feature?

File: README.md [hunk 6]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
 ./code-runner-server -port <port>
 ```
 
-## Send code files and arguments to server
-Use curl command to connect to the server and send request.
-Use multipart/form data option(-F) to provide the assignment submission file and the command-line arguments that are to be used when executing the code.
+## Code Runner as a docker image
```
Comment (pradykaushik) [inline README.md]: ## Run Code Runner inside a Docker Container.

File: README.md [hunk 7]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- A docker image of code runner is used as a base image for different language specific images.
+- Code Runner is not directly run, instead use the desired language specific image to run the assignments.
```
Comment (pradykaushik) [inline README.md]: this is confusing. Instead, we should mention that the base image is not directly used and that the image pertaining to the required assignment environment should be used.

File: README.md [hunk 8]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- Prerequisite for this is that docker engine should be installed.
```
Comment (pradykaushik) [inline README.md:31]: this should be mentioned at the very beginning of this sub-section.

File: README.md [hunk 9]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- See [instructions](https://docs.docker.com/engine/installation/) for installing docker engine on different supported platforms.
+
+Use the following docker command to run the language specific image.
```
Comment (pradykaushik) [inline README.md]: "language specific image" is not accurate as it is actually the image for the required assignment environment.

File: README.md [hunk 10]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
 ```commandline
-curl <server_ip_address:port_number> -F <compressed_filepath> -F <arg1> -F <arg2> ...
+docker run --publish <port_to_expose>:<port_to_run> <language_image_tag> -port <port_to_run>
 ```
 
+## Supported Languages
+A fixed set of languages and their versions are currently supported.
```
Comment (pradykaushik) [inline README.md]: Below is the list of supported languages and the corresponding versions.

File: README.md [hunk 11]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- gcc 7
+- g++ 7
+- python 3.7
+- java 8 & 11
+
 ## Web page to Submit Assignment
-![](docs/codeRunnerPage.png)
+Use the exposed port number to open the web page. Example - for exposed port number as 52453, hit `localhost:52453` on browser.
```
Comment (pradykaushik) [inline README.md]: [http://localhost:52453]() and not [localhost:52453]().

File: README.md [hunk 12]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+
+![](docs/webPage.png)
 
 Submit your assignment through the web page as shown in the figure above.
+- Initially `Build` and `Run` buttons are disabled. Submit the assignment details to enable them.
```
Comment (pradykaushik) [inline README.md]: what details?

File: README.md [hunk 13]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
 - Choose the assignment tar ball.
 - Enter commands to compile and run the code.
-- Enter working directory to run the commands. If not provided then commands will be run at the root assignment directory.
+- Enter working directory to run the commands. If not provided then commands will be run at the assignment root directory.
 - Add command line arguments if required to run the assignment using `Add` button.
 - Click on `Submit` button to upload the details to the server.
 - The status of upload will be displayed in the `Output` section.
-- Click on `Build & Run` button to run the code.
+- Click on `Build` button to compile the code.
+- The status of build will be displayed in the `Output` section.
```
Comment (pradykaushik) [inline README.md]: club this with the previous point.

File: README.md [hunk 14]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
 
 Submit your assignment through the web page as shown in the figure above.
+- Initially `Build` and `Run` buttons are disabled. Submit the assignment details to enable them.
 - Choose the assignment tar ball.
 - Enter commands to compile and run the code.
-- Enter working directory to run the commands. If not provided then commands will be run at the root assignment directory.
+- Enter working directory to run the commands. If not provided then commands will be run at the assignment root directory.
 - Add command line arguments if required to run the assignment using `Add` button.
 - Click on `Submit` button to upload the details to the server.
 - The status of upload will be displayed in the `Output` section.
```
Comment (pradykaushik) [inline README.md]: club this with the previous point.

File: README.md [hunk 15]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
-- Click on `Build & Run` button to run the code.
+- Click on `Build` button to compile the code.
+- The status of build will be displayed in the `Output` section.
+- Click on `Run` button to run the compiled code.
 - The output will be displayed in the `Output` section as shown above.
```
Comment (pradykaushik) [inline README.md]: club this with the previous point.

File: README.md [hunk 16]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+
+#### Add Command Line Arguments
```
Comment (pradykaushik) [inline README.md]: #### Command Line Arguments

File: README.md [hunk 17]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- Command Line Arguments can be added as key-value pairs in the text boxes.
```
Comment (pradykaushik) [inline README.md]: don't need to mention "in the text boxes"

File: README.md [hunk 18]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- First text box is for key and second is for the value.
```
Comment (pradykaushik) [inline README.md]: not required.

File: README.md [hunk 19]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
-- Click on `Build & Run` button to run the code.
+- Click on `Build` button to compile the code.
+- The status of build will be displayed in the `Output` section.
+- Click on `Run` button to run the compiled code.
 - The output will be displayed in the `Output` section as shown above.
+
+#### Add Command Line Arguments
+- Command Line Arguments can be added as key-value pairs in the text boxes.
+- First text box is for key and second is for the value. 
+- If the arguments have only value then use second text box for the value. First can be left empty.
```
Comment (pradykaushik) [inline README.md]: If specifying only the value, then use the text box on the right hand side.

File: README.md [hunk 20]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+
+
+#### Run Assignment for Interpreted Languages
```
Comment (pradykaushik) [inline README.md]: #### Interpreted Languages

File: README.md [hunk 21]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- The only supported interpreted language is python.
```
Comment (pradykaushik) [inline README.md]: Currently, the only supported interpreted language is python.

File: README.md [hunk 22]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- `Build` button remains disabled if the application being run is for any of the interpreted languages.
+- Hence instead of entering command to compile, enter only command to run.
```
Comment (pradykaushik) [inline README.md]: Directly enter command to run.

File: README.md [hunk 23]
```diff
@@ -30,22 +28,50 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+- After submitting the assignment details click on `Run` button directly to run the assignment.
```
Comment (pradykaushik) [inline README.md]: Click on `Run` button to run the assignment.

**Review by pradykaushik — CHANGES_REQUESTED** [review]
Features section requires a complete rewrite.

File: README.md (hunk 1 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 2 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (balandi1) [reply README.md]: Okay. I have made them as sub points.

File: README.md (hunk 23 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 22 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 21 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 20 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 19 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 18 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 17 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 16 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 15 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 14 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 13 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 12 above)
Comment (balandi1) [reply README.md]: Details as in the tarball, commands, working directory etc. But I guess I should simply say 'Submit the assignment to enable them.'

File: README.md (hunk 11 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 9 above)
Comment (balandi1) [reply README.md]: Agreed. Changed it.

File: README.md (hunk 8 above)
Comment (balandi1) [reply README.md:31]: Done

File: README.md (hunk 7 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (balandi1) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (balandi1) [reply README.md]: My mistake. Removed it.

File: README.md (hunk 4 above)
Comment (balandi1) [reply README.md]: Okay

File: README.md [hunk 24]
```diff
@@ -30,22 +27,45 @@ Use the `-port` option to specify the port number for the server to listen reque
...
+## Supported Languages
+Below is the list of supported languages and the corresponding versions.
+- gcc 7
+- g++ 7
+- python 3.7
+- java 8 & 11
+
 ## Web page to Submit Assignment
-![](docs/codeRunnerPage.png)
+Use the exposed port number to open the web page. Example - for exposed port number as 52453, hit `http://localhost:52453` on browser.
```
Comment (pradykaushik) [inline README.md]: Maybe make the url a link? This will help them click on it and see the application landing page.
`[http://localhost:52453]()`

**Review by pradykaushik — APPROVED** [review]
Just a minor fix and everything else looks good.

File: README.md (hunk 24 above)
Comment (balandi1) [reply README.md:49]: Sure.

Comment (balandi1) [reply README.md]: Okay. But unless the container is running, `localhost:52453` won't display anything.

Comment (balandi1) [reply README.md]: Maybe displaying the index.html on click of the link can be done - `[http://localhost:52453](https://htmlpreview.github.io/?https://github.com/assignment-exec/code-runner/blob/master/client/index.html)`

**Pull request merged**